File: spark.streaming/number_streaming.py
from sys      import argv, exit
from operator import add, sub

from pyspark           import SparkContext
from pyspark.streaming import StreamingContext

if __name__ == "__main__":
    if len(argv) != 3:
        print "Please pass host and port number as arguments."
        exit(-1)
    sc = SparkContext()
    ssc = StreamingContext(sc, 1)
    ssc.checkpoint('/tmp/sparkstreamingcheckpoint')

    lines = ssc.socketTextStream(argv[1], int(argv[2]))
    sums =                                    \
      lines                                   \
      .map( lambda line: int(line) )      \
      .map( lambda number: ('aggr', number) ) \
      .reduceByKeyAndWindow(add, sub, 30, 5)
    sums.pprint()

    ssc.start()
    ssc.awaitTermination()

File: spark.streaming/NumberStreaming.scala
import org.apache.spark._
import org.apache.spark.streaming._

object Streaming {
  def main(args: Array[String]) {
    val conf = new SparkConf().setMaster("local[*]").setAppName("streaming")
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")
    val ssc = new StreamingContext(sc, Seconds(1))
    ssc.checkpoint("/tmp/sparkstreamingcheckpoint")

    val lines = ssc.socketTextStream("localhost", 9999)

    val sums = lines
               .map( i => ("aggr", i.toInt) )
               .reduceByKeyAndWindow(_ + _, _ - _, Seconds(30), Seconds(5) )

    sums.print()

    ssc.start()
    ssc.awaitTermination()

    println("Spark Streaming from a Kafka topic")
  }
}
